src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs

```diff
@@ -0,0 +1,365 @@
+                foreach (var doc in r.Value)
+                {
+                    doc.Json.Dispose();
+                }
+            }
+        }
+
+        private class ReduceOutputReferencesSubCommand
+        {
+            private readonly MapReduceIndex _index;
+            private readonly string _outputReduceToCollection;
+            private readonly Transaction _indexWriteTransaction;
+            private readonly DocumentDatabase _database;
+            readonly Dictionary<string, HashSet<string>> _referencesOfReduceOutputs = new Dictionary<string, HashSet<string>>();
+            readonly Dictionary<string, HashSet<string>> _idsToDeleteByReferenceDocumentId = new Dictionary<string, HashSet<string>>();
+
+            public ReduceOutputReferencesSubCommand(MapReduceIndex index, string outputReduceToCollection, Transaction indexWriteTransaction)
+            {
+                _index = index;
+                _outputReduceToCollection = outputReduceToCollection;
+                _indexWriteTransaction = indexWriteTransaction;
+                _database = index.DocumentDatabase;
+            }
+
+            public void Add(string referenceDocumentId, string reduceOutputDocumentId)
+            {
+                if (_referencesOfReduceOutputs.TryGetValue(referenceDocumentId, out var values) == false)
+                {
+                    values = new HashSet<string>(1);
+
+                    _referencesOfReduceOutputs.Add(referenceDocumentId, values);
+                }
+
+                values.Add(reduceOutputDocumentId);
+            }
+
+            public void DeleteFromReferenceDocument(string reduceOutputId)
+            {
+                var referenceId = _index.ReduceOutputs.GetPatternGeneratedIdForReduceOutput(_indexWriteTransaction, reduceOutputId);
+
+                if (_idsToDeleteByReferenceDocumentId.TryGetValue(referenceId, out var values) == false)
+                {
+                    values = new HashSet<string>(1);
+
+                    _idsToDeleteByReferenceDocumentId.Add(referenceId, values);
+                }
+
+                values.Add(reduceOutputId);
+            }
+
+            public void Execute(DocumentsOperationContext context)
+            {
+                // delete
+
+                foreach (var reduceReferenceIdToReduceOutputIds in _idsToDeleteByReferenceDocumentId)
+                {
+                    var referenceDocument = _database.DocumentsStorage.Get(context, reduceReferenceIdToReduceOutputIds.Key);
+                    
+                    if (referenceDocument == null)
+                        continue;
+
+                    if (referenceDocument.Data.TryGet(nameof(ReduceOutputIdsReference.ReduceOutputs), out BlittableJsonReaderArray ids) == false)
+                        ThrowIdsPropertyNotFound(referenceDocument.Id);
+
+                    var idsToRemove = reduceReferenceIdToReduceOutputIds.Value;
+
+                    if (idsToRemove.Count >= ids.Length)
```

Comment:
Is it possible that we have ids in the `idsToRemove` that aren't in the `ids` ? 
I think we need to do the work to remove it always

Reply:
Not possible. I'll add `Debug.Assert` for that.